Core/Authserver: Auth cleanup phase 1

Treeston opened the pull request:
~~Nothing too huge yet, just some minor refactors. I'm trying to keep PR size low so they can be merged in a reasonable timeframe.~~

OK yeah nevermind it ended up getting bloated. Refactors have a tendency to lead to more refactors. Oops.

It's all building up to an eventual fix for #22913 in case that wasn't sufficiently implied.

---

## `AuthSession`
* Aggressive `std::array`-ification of fields
* Actually readable now due to aggressive use of `::GetDigestOf`
* A bunch of `BigNumber`s that were only ever used as byte arrays are now... well, byte arrays
  * In particular the session key, both here and in `WorldSession` (and related logic)

## `ARC4`
* Moved to `Trinity::Crypto`
* Normalized parameter order from `(int, uint8*)` to `(uint8*, size_t)`
* Added generic container overloads
* Moved seed length from ctor to `::Init`, I have no idea why this was in the ctor before

## `BigNumber`
* Added `BigNumber(uint32)` ctor
* Added `BigNumber(std::string const&)` ctor
* Added `SetBinary(C const&)` for generic containers
* Added `SetHexStr(std::string const&)`
* Removed all the `AsByteArray` variants
* Added `GetBytes` to retrieve bytes into a provided fixed-length buffer
* Added `ToByteArray<N>` to retrieve bytes as a fixed-size array
* Added `ToByteVector` to retrieve bytes into a vector

## `CryptoRandom`
* New header that's mostly just a wrapper around `RAND_bytes`
* `void GetRandomBytes(uint8*, size_t)` for arbitrary buffers
* `void GetRandomBytes(Container&)` for pre-sized containers
* `std::array<uint8, S> GetRandomBytes<S>()` to give you a new array

## `SessionKeyGenerator`
* The old `SHA1Randx`, used for Warden key generation
* It's now called `SessionKeyGenerator` because that's what it's called in master
* It's now templated on the hash function to use (because that's how it works in master)
* It is now aggressively more readable because of the next section

## `SHA1` & `HMAC_SHA1`
* Renamed `SHA1Hash` to `SHA1`, and `HmacHash` to `HMAC_SHA1`
* Moved both to the `Trinity::Crypto` namespace where they belong
* Both now expose a unified interface:
  * `static constexpr size_t DIGEST_LENGTH`
  * `using Digest = std::array<uint8, DIGEST_LENGTH>`
  * `void UpdateData(uint8 const*, size_t)` for arbitrary data
  * `void UpdateData(C const&)` for various container types (including `std::string`)
  * `static Digest GetDigestOf(Ts&&...)` for objectless hashing of one or more supported types

## Other changes
* `ByteArrayToHexStr` (in `Util.h`) now takes `size_t` instead of `int` for its length parameter
* `ByteArrayToHexStr` now has an overload for generic containers

Treeston committed:
stage 1

Treeston committed:
Fix a 1/256 bug in UpdateBigNumbers with msbyte zero. UpdateBigNumbers now explicitly requires the number of bytes to provide.

Treeston committed:
punt stuff to cpp

Treeston committed:
right, proper size qualifiers, i've been too spoiled working with fmtlib

Treeston committed:
kill sha1hash::getlength(), and some int -> size_t to make compilers happy

Treeston committed:
#include <array>

Warpten commented on src/common/Cryptography/SHA1.h:
```diff
@@ -29,21 +30,36 @@ class TC_COMMON_API SHA1Hash
+            if constexpr (sizeof...(Ts) > 0)
+                UpdateBigNumbers<SIZES...>(std::forward<Ts>(tail)...);
+        }
+
```
```cpp
    template <
        typename... Ts>
    auto UpdateBigNumbers(Ts&&... ts) -> std::enable_if_t<
        (std::is_same_v<foo, std::decay_t<Ts>> && ...)
    , void>
    {
        return (UpdateBigNumber(std::forward<Ts>(ts)), ...);
    }
```

You can tweak the decay bit, i was feeling lazy

Treeston replied:
asserting can produce a nicer error message though

Warpten replied:
sure. fold expressions though

Treeston replied:
ha sure we can do it

Warpten commented on src/common/Utilities/Util.cpp:
```diff
@@ -611,7 +611,7 @@ bool Utf8ToUpperOnlyLatin(std::string& utf8String)
 }
 
-std::string ByteArrayToHexStr(uint8 const* bytes, uint32 arrayLen, bool reverse /* = false */)
+std::string ByteArrayToHexStr(uint8 const* bytes, size_t arrayLen, bool reverse /* = false */)
```
cue anti-boost protests

`boost::algorithm::(un)hex`?

Treeston replied:
sure, i honestly just wanted to replace the parameter, we can change that

Warpten commented on src/common/Utilities/Util.h:
```diff
@@ -304,7 +304,8 @@ TC_COMMON_API bool IsIPAddress(char const* ipaddress);
 
-TC_COMMON_API std::string ByteArrayToHexStr(uint8 const* bytes, uint32 length, bool reverse = false);
+TC_COMMON_API std::string ByteArrayToHexStr(uint8 const* bytes, size_t length, bool reverse = false);
+template <typename C> std::string ByteArrayToHexStr(C const& container, bool reverse = false) { return ByteArrayToHexStr(std::data(container), std::size(container), reverse); }
```
TC_COMMON_API lacking here kind of defeats the purpose of this method being here (and you can't add it)

Treeston replied:
not sure i follow, `TC_*_API` shouldn't be necessary for any method defined in the header

Warpten replied:
IIRC `TC_*_API` macros are for hotswap
except you can't have declspec(dllexport) (which is what those macros do) on template methods for obvious reasons

so you broke hotswap builds

Warpten replied:
actually i don't know i just know TC_API macro breaks it here if its added and apparently it builds on nix for obscure reasons so ... ???

Treeston replied:
`TC_COMMON_API` resolves to `export` inside common, and to `import` outside common

but that only matters if the definition is only in common, if the definition is in the header then it's in every compilation unit anyway

Treeston committed:
fold expressions to make warpten a happy camper

Shauren commented on src/common/Cryptography/SHA1.h:
```diff
@@ -29,21 +30,34 @@ class TC_COMMON_API SHA1Hash
+            UpdateData(arena.data(), SIZE);
+        }
+
+        SHA_CTX _mC;
```
thats a not properly executed codestyle adjustment, old name `mC` was a variation of `m_C` without the underscore, you should have removed the m as well

Treeston committed:
more code style adjustments for shauren

Treeston committed:
kill off UpdateBigNumbers, add generic UpdateData overload for containers, replace usages

Treeston committed:
kill off some old-ass warden code, i'm pretty sure that used to be a workaround for the bug i fixed in b2b98803, heh

Treeston committed:
refactor BigNumber, AsByteArray is dead, we now have GetBytes, ToByteVector, and ToByteArray

Warpten commented on src/common/Cryptography/SHA1.h:
```diff
@@ -29,21 +30,22 @@ class TC_COMMON_API SHA1Hash
 
         void UpdateData(const uint8 *dta, int len);
         void UpdateData(const std::string &str);
+        void UpdateData(char const* str) { UpdateData(std::string(str)); } /* explicit overload to avoid using the container template */
```
<details>
<summary>std::string_view overload?</summary>

cue angry shauren noises
</details>

Treeston committed:
std::string_view overload

Shauren commented on src/common/Cryptography/WardenKeyGeneration.h:
```diff
@@ -25,18 +25,16 @@
 {
 public:
-    SHA1Randx(uint8* buff, uint32 size)
+    SHA1Randx(std::array<uint8, 40> const& buf)
```
Size of this array isnt fixed and taking half of it in code below is the correct behavior, see master branch: https://github.com/TrinityCore/TrinityCore/blob/8f9a69c9fa812b82af97809e445fc176b28f90eb/src/common/Cryptography/SessionKeyGeneration.h#L28 
It was renamed and is being used with two different sizes right now, 40 and 32

Treeston committed:
Add CryptoRandom.cpp + .h so we can stop having random (heh) RAND_bytes calls all over the place.

Unify hash function class layouts (SHA1 + HMACSHA1).
Move WardenKeyGeneration.h -> SessionKeyGenerator.hpp (matches master).
Unfuck ARC4 and AuthCrypt.
Move everything into the Trinity::Crypto namespace.

Treeston committed:
dear lord AuthSession.cpp is starting to look READABLE what is this sorcery

Treeston committed:
it's a commit that removes a single whitespace from a file to make circle-ci not throw a hissy fit

Treeston committed:
more CI babysitting

Treeston committed:
more CI-related cleanup

Treeston committed:
headers

Treeston committed:
just in case

Carbenium commented:
@Treeston Just a side-note since you are touching the crypto stuff: The OpenSSL low-level API (`SHA*_` and so on) is being officially deprecated with OpenSSL 3.0 (https://wiki.openssl.org/index.php/OpenSSL_3.0#Low_Level_APIs).
Might be a good time to switch to the `EVP_*` family of functions.

Shauren commented on src/common/Cryptography/HMACSHA1.h:
```diff
@@ -19,27 +19,76 @@
+{
+namespace Crypto
+{
+class TC_COMMON_API HMAC_SHA1
```
I don't like how this gets a refactor going in a different direction than in master branch

https://github.com/TrinityCore/TrinityCore/blob/b3ba029791a9003880d77a048b16e9e593485fe9/src/common/Cryptography/HmacHash.h#L49-L50

Treeston committed:
Make hash/HMAC logic generic, SHA1/HMAC_SHA1 are now just instantiations of that. Update SHA1 to use openssl EVP.

Carbenium commented:
@Shauren Can we drop OpenSSL 1.0 support? Even Debian Oldstable (Stretch) is on 1.1

Carbenium commented on src/common/Cryptography/HMAC.h:
```diff
@@ -0,0 +1,138 @@
+#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER < 0x10100000L
+            static HMAC_CTX* MakeCTX()
+            {
+                HMAC_CTX* ctx = new HMAC_CTX();
```
https://www.openssl.org/docs/manmaster/man3/EVP_DigestSignInit.html

These are the corresponding `EVP_*` functions.

Treeston replied:
Can you find any documentation on the `PKEY` functions for openssl < 1.1.1? I couldn't.

Carbenium replied:
Should be solved when we drop <1.1.1.

Carbenium commented on src/common/Cryptography/SHA1.h:
```diff
@@ -15,60 +15,18 @@
-#include <string>
-#include <string_view>
+#include "GenericHash.h"
 #include <openssl/sha.h>
```
```suggestion
#include <openssl/evp.h>
```

Carbenium commented on src/common/Cryptography/HMAC_SHA1.h:
```diff
@@ -15,23 +15,18 @@
-    SHA1_Init(&_ctx);
-}
+#include "HMAC.h"
+#include <openssl/sha.h>
```
```suggestion
#include <openssl/evp.h>
```

Carbenium replied:
Actually there is one problem: `SHA_DIGEST_LENGTH` is in `<openssl/sha.h>`
Couple of options:
  * keep `sha.h` (not in favour)
  * define it ourselves
  * use `EVP_MD_size/EVP_MD_CTX_size`

Treeston replied:
The third option isn't an option, these aren't `constexpr`.

Treeston committed:
Kill HMAC_SHA1.h and SHA1.h. Rename GenericHash.h to Hash.h, now includes the SHA1 and SHA256 typedefs. HMAC.h now includes HMAC_SHA1 and HMAC_SHA256 typedefs. New CryptoConstants.h.

Treeston committed:
Hash.h -> CryptoHash.h to avoid a clash

Treeston committed:
string_view as value, and remove a stray auto

Treeston committed:
get rid of two undesirable and unused overloads (use containers, kids)

Carbenium commented on src/common/Cryptography/ARC4.cpp:
```diff
@@ -17,34 +17,26 @@
-    EVP_EncryptInit_ex(m_ctx, EVP_rc4(), nullptr, nullptr, nullptr);
-    EVP_CIPHER_CTX_set_key_length(m_ctx, len);
+    EVP_CIPHER_CTX_init(_ctx);
+    EVP_EncryptInit_ex(_ctx, EVP_rc4(), nullptr, nullptr, nullptr);
```
Maybe some asserts sprinkled in here wouldn't be the worst idea.

Treeston replied:
ah yes, i missed those

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,111 @@
+            typedef EVP_MD const* (*HashCreator)();
+
+#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER < 0x10100000L
+            static EVP_MD_CTX* MakeCTX() { return EVP_MD_CTX_create(); }
```
same here.

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,111 @@
+
+#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER < 0x10100000L
+            static EVP_MD_CTX* MakeCTX() { return EVP_MD_CTX_create(); }
+            void DestroyCTX(EVP_MD_CTX* ctx) { EVP_MD_CTX_destroy(ctx); }
```
I don't see any calls to `GenericHashImpl::DestroyCTX`

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,111 @@
+                    int result = EVP_DigestFinal_ex(_ctx, _digest.data(), &length);
+                    ASSERT(result == 1);
+                    ASSERT(length == DIGEST_LENGTH);
+                    _ctx = nullptr;
```
Either call `EVP_MD_CTX_free` or reuse the same ctx (has to first inited with `EVP_DigestInit_ex` again).

Treeston committed:
add various asserts (and a destructor i missed, gdi this is why we have smart pointers)

Treeston committed:
this is exactly why we have smart pointers

Treeston committed:
tfw assert isn't even included yet

Carbenium commented on src/common/Cryptography/BigNumber.cpp:
```diff
@@ -185,15 +179,14 @@ bool BigNumber::IsNegative() const
 {
     int nBytes = GetNumBytes();
-    ASSERT(!(nBytes < 0));
+    ASSERT(!(nBytes < 0), "Bignum has negative number of bytes (%d).", nBytes);
```
```suggestion
    ASSERT(nBytes >= 0, "Bignum has negative number of bytes (%d).", nBytes);
```
if you like. I find it more readable.

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,120 @@
+                static constexpr size_t DIGEST_LENGTH = DigestLength;
+                using Digest = std::array<uint8, DIGEST_LENGTH>;
+
+                static Digest GetDigestOf(uint8 const* dta, int len)
```
```suggestion
                static Digest GetDigestOf(uint8 const* data, size_t len)
```

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,120 @@
+                static Digest GetDigestOf(uint8 const* dta, int len)
+                {
+                    GenericHash hash;
+                    hash.UpdateData(dta, len);
```
```suggestion
                    hash.UpdateData(data, len);
```

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,120 @@
+                    _ctx = nullptr;
+                }
+
+                void UpdateData(uint8 const* dta, size_t len)
```
```suggestion
                void UpdateData(uint8 const* data, size_t len)
```

Carbenium commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,120 @@
+
+                void UpdateData(uint8 const* dta, size_t len)
+                {
+                    int result = EVP_DigestUpdate(_ctx, dta, len);
```
```suggestion
                    int result = EVP_DigestUpdate(_ctx, data, len);
```

Warpten commented on src/common/Cryptography/HMAC.h:
```diff
@@ -0,0 +1,132 @@
+            static HMAC_CTX* MakeCTX()
+            {
+                HMAC_CTX* ctx = new HMAC_CTX();
+                int result = HMAC_CTX_init(ctx);
```
```suggestion
                [[maybe_unused]] int result = HMAC_CTX_init(ctx);
```

keeps the linters happy

Treeston replied:
no that one actually doesn't have a return value, that's why appveyor died

(smh openssl 1.0.2)

Warpten commented on src/common/Cryptography/CryptoHash.h:
```diff
@@ -0,0 +1,120 @@
+    namespace Crypto
+    {
+        using SHA1 = Trinity::Impl::GenericHash<EVP_sha1, Constants::SHA1_DIGEST_LENGTH_BYTES>;
+        using SHA256 = Trinity::Impl::GenericHash<EVP_sha256, Constants::SHA256_DIGEST_LENGTH_BYTES>;
```
This is vastly irrelevant but i'd use openssl defines here since you're already using openssl types, just for consistency

Carbenium replied:
We discussed this previously. See https://github.com/TrinityCore/TrinityCore/pull/25093#discussion_r460142261 . 
That's why we ended up here.

Warpten replied:
Fair

Carbenium commented:
Alright, if Warpten is even going for linter warnings, then I have also another nit-pick ;)

How about C++17 syntax for nested namespaces?
```c++
namespace Trinity::Crypto
{
...
```

Treeston committed:
more cleanup, get rid of a bunch of indentations by using nested namespaces

Treeston committed:
clean up some shady casts

Carbenium reviewed (approved):
Just some more nit-picks ;) Otherwise LGTM!

Treeston committed:
code style tweaks

Treeston committed:
more cleanups

Carbenium commented on src/server/game/Server/WorldSocket.cpp:
```diff
@@ -501,17 +498,17 @@ void WorldSocket::HandleAuthSessionCallback(std::shared_ptr<AuthSession> authSes
 
     // Check that Key and account name are the same on client and server
-    uint32 t = 0;
+    uint8 t[4] = { 0x00,0x00,0x00,0x00 };
```
no love for whitespaces? :cry:

Warpten commented:
missed a few more of these `[[maybe_unused]]` on ossl result codes

Shauren commented on src/common/Cryptography/HMAC.h:
```diff
@@ -0,0 +1,128 @@
+                int result = HMAC_Init_ex(_ctx, seed, len, HashCreator(), nullptr);
+                ASSERT(result == 1);
+            }
+            template <typename C> GenericHMAC(C const& container) : GenericHMAC(std::data(container), std::size(container)) {}
```
Why are you placing the template part on the same line as everything else? We don't do that anywhere

Treeston committed:
final cleanup

illfated commented:
Great job. Love your style.
